db.go

```diff
@@ -632,6 +654,10 @@ func (d *DB) Close() error {
 
 // Compact the specified range of keys in the database.
 func (d *DB) Compact(start, end []byte /* CompactionOptions */) error {
+	if atomic.LoadInt32(&d.closed) != 0 {
```

Comment:
I'm interested in the pebble work you guys are doing and also learning go.  Excuse me if I'm completely wrong but it seems that since these guard clauses aren't inside the `mu.Lock(), code subsequent to them may be executed even if the DB is closed (very low chance but still possible).  For example, in this case the `atomic.LoadInt32(&d.closed) != 0` could be false but straight after `atomic.StoreInt32(&d.closed, 1)` could be called within the `Close()` function, then this function would try to get the `mu.Lock()` and would block until the `Close()` finished and then execute the remainder of the `Compact` code not realizing the DB was already closed.  To be honest, I don't understand the compact code well enough to know whether this would cause problems in the compact or would just mean the `ErrClosed` would not be reported (which might be fine).

I won't make this an actual review because I don't have the knowledge of your code base to actually review the code but I thought I would raise this anyway.  Is this a potential issue?

Reply:
@damienhollis You're correct that performing these checks outside of `d.mu.Lock` makes them inaccurate, but I don't think it is possible to make the checks perfectly accurate even with different locking. The root problem is that if you're calling `Close` concurrently with any other option your program is in error. These guard rails are simply trying to give you early warning about that incorrect usage. Similar to the way that Go's builtin map tries to detect concurrent access. It can't do it perfectly, but tries instead to do something cheaply. 

I wonder if I should replace the errors with panics to make it more explicit that these are unexpected conditions.

Reply:
@petermattis thanks for the reply, makes sense.